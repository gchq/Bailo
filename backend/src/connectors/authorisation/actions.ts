import { TokenActions } from '../../models/Token.js'

export const ModelAction = {
  Create: 'model:create',
  View: 'model:view',
  Update: 'model:update',
  Write: 'model:write',
} as const
export type ModelActionKeys = (typeof ModelAction)[keyof typeof ModelAction]

export const ReleaseAction = {
  Create: 'release:create',
  View: 'release:view',
  Delete: 'release:delete',
  Update: 'release:update',
} as const
export type ReleaseActionKeys = (typeof ReleaseAction)[keyof typeof ReleaseAction]

export const AccessRequestAction = {
  Create: 'access_request:create',
  View: 'access_request:view',
  Update: 'access_request:update',
  Delete: 'access_request:delete',
} as const
export type AccessRequestActionKeys = (typeof AccessRequestAction)[keyof typeof AccessRequestAction]

export const SchemaAction = {
  Create: 'schema:create',
  Delete: 'schema:delete',
  Update: 'schema:update',
} as const
export type SchemaActionKeys = (typeof SchemaAction)[keyof typeof SchemaAction]

export const FileAction = {
  Delete: 'file:delete',
  Upload: 'file:upload',
  // 'view' refers to the ability to see metadata about the file.  'download' lets the user view the file contents.
  View: 'file:view',
  Download: 'file:download',
  Update: 'file:update',
} as const
export type FileActionKeys = (typeof FileAction)[keyof typeof FileAction]

export const ImageAction = {
  Pull: 'image:pull',
  Push: 'image:push',
  List: 'image:list',
  Wildcard: 'image:wildcard',
  Delete: 'image:delete',
} as const
export type ImageActionKeys = (typeof ImageAction)[keyof typeof ImageAction]

export const ResponseAction = {
  Create: 'response:create',
  View: 'response:view',
  Update: 'response:update',
} as const
export type ResponseActionKeys = (typeof ResponseAction)[keyof typeof ResponseAction]

export const ActionLookup = {
  [ModelAction.Create]: TokenActions.ModelWrite.id,
  [ModelAction.View]: TokenActions.ModelRead.id,
  [ModelAction.Update]: TokenActions.ModelWrite.id,
  [ModelAction.Write]: TokenActions.ModelWrite.id,

  [ReleaseAction.Create]: TokenActions.ReleaseWrite.id,
  [ReleaseAction.View]: TokenActions.ReleaseRead.id,
  [ReleaseAction.Delete]: TokenActions.ReleaseWrite.id,
  [ReleaseAction.Update]: TokenActions.ReleaseWrite.id,

  [AccessRequestAction.Create]: TokenActions.AccessRequestWrite.id,
  [AccessRequestAction.View]: TokenActions.AccessRequestRead.id,
  [AccessRequestAction.Update]: TokenActions.AccessRequestWrite.id,
  [AccessRequestAction.Delete]: TokenActions.AccessRequestWrite.id,

  [SchemaAction.Create]: TokenActions.SchemaWrite.id,
  [SchemaAction.Delete]: TokenActions.SchemaWrite.id,
  [SchemaAction.Update]: TokenActions.SchemaWrite.id,

  [FileAction.Delete]: TokenActions.FileWrite.id,
  [FileAction.Upload]: TokenActions.FileWrite.id,
  [FileAction.View]: TokenActions.FileRead.id,
  [FileAction.Download]: TokenActions.FileRead.id,
  [FileAction.Update]: TokenActions.FileWrite.id,

  [ImageAction.Pull]: TokenActions.ImageRead.id,
  [ImageAction.Push]: TokenActions.ImageWrite.id,
  [ImageAction.List]: TokenActions.ImageRead.id,
  [ImageAction.Wildcard]: TokenActions.ImageWrite.id,
  [ImageAction.Delete]: TokenActions.ImageWrite.id,
} as const
export type ActionLookupKeys = (typeof ActionLookup)[keyof typeof ActionLookup]
