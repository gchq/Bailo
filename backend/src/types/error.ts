import Logger from 'bunyan'

export interface BailoError extends Error {
  // Inherited from 'Error'
  // name: string
  // message: string

  // An HTTP response code that represents the error
  name: string
  message: string
  code: number
  status: number

  // This data is logged publicly to the frontend
  context?: {
    documentationUrl?: string

    // Items in internal are not displayed to the user and may include sensitive information
    internal?: unknown

    [key: string]: unknown
  }

  // A custom logger may be provided, otherwise a default is used
  logger?: Logger
}

export function isBailoError(err: unknown): err is BailoError {
  if (typeof err !== 'object' || err === null) {
    return false
  }

  if (err instanceof Error && err.name === 'Bailo Error') {
    return true
  }

  return false
}
