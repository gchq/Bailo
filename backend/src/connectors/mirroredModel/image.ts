import { json } from 'node:stream/consumers'

import { PassThrough } from 'stream'
import { finished } from 'stream/promises'
import { Headers } from 'tar-stream'

import { ModelDoc } from '../../models/Model.js'
import { ReleaseDoc } from '../../models/Release.js'
import { UserInterface } from '../../models/User.js'
import log from '../../services/log.js'
import { addCompressedRegistryImageComponents } from '../../services/mirroredModel/mirroredModel.js'
import { initialiseTarGzUpload } from '../../services/mirroredModel/tarball.js'
import {
  doesImageLayerExist,
  initialiseImageUpload,
  joinDistributionPackageName,
  putImageBlob,
  putImageManifest,
  splitDistributionPackageName,
} from '../../services/registry.js'
import { getReleaseBySemver } from '../../services/release.js'
import config from '../../utils/config.js'
import { Forbidden, InternalError } from '../../utils/error.js'
import { hasKeysOfType } from '../../utils/typeguards.js'
import { ReleaseAction } from '../authorisation/actions.js'
import authorisation from '../authorisation/index.js'
import { BaseExporter, BaseImporter, BaseMirrorMetadata, requiresInit } from './base.js'
import { MirrorKind, MirrorKindKeys } from './index.js'

export type ImageMirrorMetadata = BaseMirrorMetadata & {
  importKind: MirrorKindKeys<'Image'>
  distributionPackageName: string
}
export type ImageMirrorInformation = {
  metadata: ImageMirrorMetadata
  image: { modelId: string; imageName: string; imageTag: string }
}

export class ImageExporter extends BaseExporter {
  protected readonly semver: string
  protected release: ReleaseDoc | undefined
  protected readonly imageId: string
  protected image: ReleaseDoc['images'][number] | undefined
  protected distributionPackageName: string | undefined

  constructor(
    user: UserInterface,
    model: string | ModelDoc,
    release: string | ReleaseDoc,
    image: string | ReleaseDoc['images'][number],
    logData?: Record<string, unknown>,
  ) {
    super(user, model, logData)

    if (typeof release === 'string') {
      this.semver = release
    } else {
      this.release = release
      this.semver = this.release.semver
    }
    if (typeof image === 'string') {
      this.imageId = image
    } else {
      this.image = image
      this.imageId = this.image._id.toString()
    }
  }

  getRelease() {
    return this.release
  }

  getImage() {
    return this.image
  }

  getDistributionPackageName() {
    return this.distributionPackageName
  }

  async _init() {
    await super._init()

    if (!this.release) {
      this.release = await getReleaseBySemver(this.user, this.model!, this.semver)
    }
    const releaseAuth = await authorisation.release(this.user, this.model!, ReleaseAction.View, this.release)
    if (!releaseAuth.success) {
      throw Forbidden(releaseAuth.info, {
        userDn: this.user.dn,
        modelId: this.modelId,
        semver: this.semver,
        ...this.logData,
      })
    }

    if (!this.image) {
      this.image = this.release.images.find((image) => image._id.toString() === this.imageId)
      if (!this.image) {
        throw InternalError('Could not find image associated with release.', {
          modelId: this.modelId,
          semver: this.semver,
          imageId: this.imageId,
          ...this.logData,
        })
      }
    }
    const imageAuth = await authorisation.image(this.user, this.model!, {
      type: 'repository',
      name: this.modelId,
      actions: ['pull'],
    })
    if (!imageAuth.success) {
      throw Forbidden(imageAuth.info, {
        userDn: this.user.dn,
        modelId: this.modelId,
        semver: this.semver,
        imageId: this.imageId,
        ...this.logData,
      })
    }

    // update the distributionPackageName to use the mirroredModelId
    const modelIdRe = new RegExp(String.raw`^${this.modelId}`)
    this.distributionPackageName = joinDistributionPackageName({
      domain: '',
      path: this.image.name.replace(modelIdRe, this.model!.settings.mirror.destinationModelId!),
      tag: this.image.tag,
    })
  }

  protected getInitialiseTarGzUploadParams(): Parameters<typeof initialiseTarGzUpload> {
    if (!this.model) {
      throw InternalError('Method `getInitialiseTarGzUploadParams` called before `this.model` defined.', {
        ...this.logData,
      })
    }
    if (!this.distributionPackageName) {
      throw InternalError(
        'Method `getInitialiseTarGzUploadParams` called before `this.distributionPackageName` defined.',
        {
          ...this.logData,
        },
      )
    }
    return [
      `${this.imageId}.tar.gz`,
      {
        exporter: this.user.dn,
        sourceModelId: this.modelId,
        mirroredModelId: this.model!.settings.mirror.destinationModelId!,
        distributionPackageName: this.distributionPackageName!,
        importKind: MirrorKind.Image,
      },
      this.logData,
    ]
  }

  @requiresInit
  async addData() {
    // Non-null assertion operator used due to `requiresInit` performing assertion
    await addCompressedRegistryImageComponents(
      this.user,
      this.modelId,
      this.distributionPackageName!,
      this.tarStream!,
      this.logData,
    )
  }
}

export class ImageImporter extends BaseImporter {
  declare protected metadata: ImageMirrorMetadata

  protected user: UserInterface
  protected imageName: string
  protected imageTag: string
  protected manifestBody: unknown = null

  protected manifestRegex = new RegExp(String.raw`^${config.modelMirror.contentDirectory}/manifest\.json$`)
  protected blobRegex = new RegExp(String.raw`^${config.modelMirror.contentDirectory}/blobs\/sha256\/[0-9a-f]{64}$`)

  constructor(user: UserInterface, metadata: ImageMirrorMetadata, logData?: Record<string, unknown>) {
    super(metadata, logData)
    if (this.metadata.importKind !== MirrorKind.Image) {
      throw InternalError('Cannot parse compressed Image: incorrect metadata specified.', {
        metadata: this.metadata,
        ...this.logData,
      })
    }

    this.user = user
    const distributionPackageNameObject = splitDistributionPackageName(this.metadata.distributionPackageName)
    if (!('tag' in distributionPackageNameObject)) {
      throw InternalError('Distribution Package Name must include a tag.', {
        distributionPackageNameObject,
        distributionPackageName: this.metadata.distributionPackageName,
      })
    }
    ;({ path: this.imageName, tag: this.imageTag } = distributionPackageNameObject)
  }

  async processEntry(entry: Headers, stream: PassThrough) {
    if (entry.type === 'file') {
      // Process file
      if (this.manifestRegex.test(entry.name)) {
        // manifest.json must be uploaded after the other layers otherwise the registry will error as the referenced layers won't yet exist
        log.debug({ ...this.logData }, 'Extracting un-tarred manifest.')
        this.manifestBody = await json(stream)
      } else if (this.blobRegex.test(entry.name)) {
        // convert filename to digest format
        const layerDigest = `${entry.name.replace(new RegExp(String.raw`^(${config.modelMirror.contentDirectory}/blobs\/sha256\/)`), 'sha256:')}`
        try {
          if (await doesImageLayerExist(this.user, this.metadata.mirroredModelId, this.imageName, layerDigest)) {
            log.debug(
              {
                name: entry.name,
                size: entry.size,
                ...this.logData,
              },
              'Skipping blob as it already exists in the registry.',
            )

            // auto-drain the stream
            stream.resume()
          } else {
            log.debug(
              {
                name: entry.name,
                size: entry.size,
                ...this.logData,
              },
              'Initiating un-tarred blob upload.',
            )
            const res = await initialiseImageUpload(this.user, this.metadata.mirroredModelId, this.imageName)

            log.debug(
              {
                name: entry.name,
                size: entry.size,
                ...this.logData,
              },
              'Putting image blob.',
            )
            await putImageBlob(
              this.user,
              this.metadata.mirroredModelId,
              this.imageName,
              res.location,
              layerDigest,
              stream,
              String(entry.size),
            )
            await finished(stream)
          }
        } catch (err) {
          throw InternalError('Failed to upload blob to registry.', {
            err,
            name: entry.name,
            size: entry.size,
            ...this.logData,
          })
        }
      } else {
        throw InternalError('Cannot parse compressed image: unrecognised contents.', { ...this.logData })
      }
    } else {
      // skip entry of type: link | symlink | directory | block-device | character-device | fifo | contiguous-file
      log.warn({ name: entry.name, type: entry.type, ...this.logData }, 'Skipping non-file entry.')
    }
  }

  async finishListener(
    resolve: (reason?: ImageMirrorInformation) => void,
    reject: (reason?: unknown) => void,
    _logData?: Record<string, unknown>,
  ) {
    log.debug({ ...this.logData }, 'Uploading manifest.')
    if (hasKeysOfType<{ mediaType: 'string' }>(this.manifestBody, { mediaType: 'string' })) {
      await putImageManifest(
        this.user,
        this.metadata.mirroredModelId,
        this.imageName,
        this.imageTag,
        JSON.stringify(this.manifestBody),
        this.manifestBody['mediaType'],
      )
      log.debug(
        {
          image: { modelId: this.metadata.mirroredModelId, imageName: this.imageName, imageTag: this.imageTag },
          ...this.logData,
        },
        'Completed registry upload',
      )

      resolve({
        metadata: this.metadata,
        image: { modelId: this.metadata.mirroredModelId, imageName: this.imageName, imageTag: this.imageTag },
      })
    } else {
      reject(InternalError('Manifest file (manifest.json) missing or invalid in Tarball file.'))
    }
  }
}
