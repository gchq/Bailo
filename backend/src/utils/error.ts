import Logger from 'bunyan'

import { RegistryErrorResponse } from '../clients/registryResponses.js'
import { BailoError } from '../types/error.js'
import { RegistryError } from '../types/RegistryError.js'

export function GenericError(code: number, message: string, context?: BailoError['context'], logger?: Logger) {
  const err = Error(message) as BailoError

  err.name = 'Bailo Error'

  if (err.stack) {
    // Remove this file from stack traces.
    err.stack = err.stack
      .split('\n')
      .filter((line) => !line.includes('backend/src/utils/v2/error.ts'))
      .join('\n')
  }

  err.code = code
  err.context = context

  Object.defineProperty(err, 'logger', {
    value: logger,
    enumerable: false,
  })

  Object.defineProperty(err, 'message', {
    value: message,
    enumerable: true,
  })

  return err
}

export function BadReq(message: string, context?: BailoError['context'], logger?: Logger) {
  return GenericError(400, message, context, logger)
}

export function Unauthorized(message: string, context?: BailoError['context'], logger?: Logger) {
  return GenericError(401, message, context, logger)
}

export function Forbidden(message: string, context?: BailoError['context'], logger?: Logger) {
  return GenericError(403, message, context, logger)
}

export function NotFound(message: string, context?: BailoError['context'], logger?: Logger) {
  return GenericError(404, message, context, logger)
}

export function ContentTooLarge(message: string, context?: BailoError['context'], logger?: Logger) {
  return GenericError(413, message, context, logger)
}

export function UnsatisfiableRange(message: string, context?: BailoError['context'], logger?: Logger) {
  return GenericError(416, message, context, logger)
}

export function InternalError(message: string, context?: BailoError['context'], logger?: Logger) {
  return GenericError(500, message, context, logger)
}

export function NotImplemented(message: string, context?: BailoError['context'], logger?: Logger) {
  return GenericError(501, message, context, logger)
}

export function ConfigurationError(message: string, context?: BailoError['context'], logger?: Logger) {
  return GenericError(503, `BAILO configuration error: ${message}`, context, logger)
}

export function RegistryError(error: RegistryErrorResponse, context: BailoError['context']) {
  const registryError = GenericError(500, `Error response received from registry.`, {
    errors: error.errors,
    ...context,
  }) as RegistryError
  registryError.name = 'Registry Error'
  registryError.errors = error.errors

  return registryError
}
