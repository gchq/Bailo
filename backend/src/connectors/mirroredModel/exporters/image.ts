import { ModelDoc } from '../../../models/Model.js'
import { ReleaseDoc } from '../../../models/Release.js'
import { UserInterface } from '../../../models/User.js'
import { MirrorLogData } from '../../../services/mirroredModel/mirroredModel.js'
import { addCompressedRegistryImageComponents } from '../../../services/mirroredModel/mirroredModel.js'
import { initialiseTarGzUpload } from '../../../services/mirroredModel/tarball.js'
import { joinDistributionPackageName } from '../../../services/registry.js'
import { Forbidden, InternalError } from '../../../utils/error.js'
import { ReleaseAction } from '../../authorisation/actions.js'
import authorisation from '../../authorisation/index.js'
import { MirrorKind } from '../index.js'
import { BaseExporter, requiresInit } from './base.js'

export class ImageExporter extends BaseExporter {
  protected readonly release: ReleaseDoc
  protected readonly image: ReleaseDoc['images'][number]
  protected distributionPackageName: string | undefined

  constructor(
    user: UserInterface,
    model: ModelDoc,
    release: ReleaseDoc,
    image: ReleaseDoc['images'][number],
    logData: MirrorLogData,
  ) {
    super(user, model, logData)
    this.release = release
    this.image = image
  }

  getRelease() {
    return this.release
  }

  getImage() {
    return this.image
  }

  getDistributionPackageName() {
    return this.distributionPackageName
  }

  protected async _init() {
    await super._init()

    const releaseAuth = await authorisation.release(this.user, this.model!, ReleaseAction.View, this.release)
    if (!releaseAuth.success) {
      throw Forbidden(releaseAuth.info, {
        userDn: this.user.dn,
        modelId: this.model.id,
        semver: this.release.semver,
        ...this.logData,
      })
    }

    const imageCheck = this.release.images.find((image) => image._id.toString() === this.image._id.toString())
    if (!imageCheck) {
      throw InternalError('Could not find image associated with release.', {
        modelId: this.model.id,
        semver: this.release.semver,
        imageId: this.image._id.toString(),
        ...this.logData,
      })
    }
    const imageAuth = await authorisation.image(this.user, this.model!, {
      type: 'repository',
      name: this.model.id,
      actions: ['pull'],
    })
    if (!imageAuth.success) {
      throw Forbidden(imageAuth.info, {
        userDn: this.user.dn,
        modelId: this.model.id,
        semver: this.release.semver,
        imageId: this.image._id.toString(),
        ...this.logData,
      })
    }

    // update the distributionPackageName to use the mirroredModelId
    const modelIdRe = new RegExp(String.raw`^${this.model.id}`)
    this.distributionPackageName = joinDistributionPackageName({
      domain: '',
      path: this.image.name.replace(modelIdRe, this.model!.settings.mirror.destinationModelId!),
      tag: this.image.tag,
    })
  }

  protected getInitialiseTarGzUploadParams(): Parameters<typeof initialiseTarGzUpload> {
    if (!this.model) {
      throw InternalError('Method `getInitialiseTarGzUploadParams` called before `this.model` defined.', {
        ...this.logData,
      })
    }
    if (!this.distributionPackageName) {
      throw InternalError(
        'Method `getInitialiseTarGzUploadParams` called before `this.distributionPackageName` defined.',
        {
          ...this.logData,
        },
      )
    }
    return [
      `${this.image._id.toString()}.tar.gz`,
      {
        schemaVersion: 1,
        exporter: this.user.dn,
        sourceModelId: this.model.id,
        mirroredModelId: this.model!.settings.mirror.destinationModelId!,
        distributionPackageName: this.distributionPackageName!,
        importKind: MirrorKind.Image,
      },
      this.logData,
    ]
  }

  @requiresInit
  async addData() {
    // Non-null assertion operator used due to `requiresInit` performing assertion
    await addCompressedRegistryImageComponents(
      this.user,
      this.model.id,
      this.distributionPackageName!,
      this.tarStream!,
      this.logData,
    )
  }
}
