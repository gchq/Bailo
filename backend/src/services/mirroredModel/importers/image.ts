import { PassThrough } from 'node:stream'
import { json } from 'node:stream/consumers'

import { escapeRegExp } from 'lodash-es'
import { finished } from 'stream/promises'
import { Headers } from 'tar-stream'

import { UserInterface } from '../../../models/User.js'
import config from '../../../utils/config.js'
import { InternalError } from '../../../utils/error.js'
import { hasKeysOfType } from '../../../utils/typeguards.js'
import log from '../../log.js'
import {
  doesImageLayerExist,
  initialiseImageUpload,
  putImageBlob,
  putImageManifest,
  splitDistributionPackageName,
} from '../../registry.js'
import { MirrorKind, MirrorKindKeys } from '../index.js'
import { MirrorLogData } from '../mirroredModel.js'
import { BaseImporter, BaseMirrorMetadata } from './base.js'

export type ImageMirrorMetadata = BaseMirrorMetadata & {
  importKind: MirrorKindKeys<'Image'>
  distributionPackageName: string
}
export type ImageMirrorInformation = {
  metadata: ImageMirrorMetadata
  image: { modelId: string; imageName: string; imageTag: string }
}

export class ImageImporter extends BaseImporter {
  declare protected readonly metadata: ImageMirrorMetadata

  protected readonly user: UserInterface
  protected readonly imageName: string
  protected readonly imageTag: string
  protected manifestBody: unknown = null

  static readonly manifestRegex = new RegExp(
    String.raw`^${escapeRegExp(config.modelMirror.contentDirectory)}/manifest\.json$`,
  )
  static readonly blobRegex = new RegExp(
    String.raw`^${escapeRegExp(config.modelMirror.contentDirectory)}/blobs\/sha256\/[0-9a-f]{64}$`,
  )

  constructor(user: UserInterface, metadata: ImageMirrorMetadata, logData: MirrorLogData) {
    super(metadata, logData)
    if (this.metadata.importKind !== MirrorKind.Image) {
      throw InternalError('Cannot parse compressed Image: incorrect metadata specified.', {
        metadata: this.metadata,
        ...this.logData,
      })
    }

    this.user = user
    const distributionPackageNameObject = splitDistributionPackageName(this.metadata.distributionPackageName)
    if (!('tag' in distributionPackageNameObject)) {
      throw InternalError('Distribution Package Name must include a tag.', {
        distributionPackageNameObject,
        distributionPackageName: this.metadata.distributionPackageName,
        metadata: this.metadata,
        ...this.logData,
      })
    }
    ;({ path: this.imageName, tag: this.imageTag } = distributionPackageNameObject)
  }

  async processEntry(entry: Headers, stream: PassThrough) {
    if (entry.type === 'file') {
      // Process file
      if (ImageImporter.manifestRegex.test(entry.name)) {
        // manifest.json must be uploaded after the other layers otherwise the registry will error as the referenced layers won't yet exist
        log.debug({ ...this.logData }, 'Extracting un-tarred manifest.')
        this.manifestBody = await json(stream)
      } else if (ImageImporter.blobRegex.test(entry.name)) {
        // convert filename to digest format
        const layerDigest = `${entry.name.replace(new RegExp(String.raw`^(${config.modelMirror.contentDirectory}/blobs\/sha256\/)`), 'sha256:')}`
        try {
          if (await doesImageLayerExist(this.user, this.metadata.mirroredModelId, this.imageName, layerDigest)) {
            log.debug(
              {
                name: entry.name,
                size: entry.size,
                ...this.logData,
              },
              'Skipping blob as it already exists in the registry.',
            )

            // auto-drain the stream
            stream.resume()
          } else {
            log.debug(
              {
                name: entry.name,
                size: entry.size,
                ...this.logData,
              },
              'Initiating un-tarred blob upload.',
            )
            const res = await initialiseImageUpload(this.user, this.metadata.mirroredModelId, this.imageName)

            log.debug(
              {
                name: entry.name,
                size: entry.size,
                ...this.logData,
              },
              'Putting image blob.',
            )
            await putImageBlob(
              this.user,
              this.metadata.mirroredModelId,
              this.imageName,
              res.location,
              layerDigest,
              stream,
              String(entry.size),
            )
            await finished(stream)
          }
        } catch (err) {
          throw InternalError('Failed to upload blob to registry.', {
            err,
            name: entry.name,
            size: entry.size,
            metadata: this.metadata,
            ...this.logData,
          })
        }
      } else {
        throw InternalError('Cannot parse compressed image: unrecognised contents.', {
          metadata: this.metadata,
          ...this.logData,
        })
      }
    } else {
      // skip entry of type: link | symlink | directory | block-device | character-device | fifo | contiguous-file
      log.warn({ name: entry.name, type: entry.type, ...this.logData }, 'Skipping non-file entry.')
    }
  }

  async finishListener(resolve: (reason?: ImageMirrorInformation) => void, reject: (reason?: unknown) => void) {
    log.debug({ ...this.logData }, 'Uploading manifest.')
    if (hasKeysOfType<{ mediaType: 'string' }>(this.manifestBody, { mediaType: 'string' })) {
      await putImageManifest(
        this.user,
        this.metadata.mirroredModelId,
        this.imageName,
        this.imageTag,
        JSON.stringify(this.manifestBody),
        this.manifestBody['mediaType'],
      )
      log.debug(
        {
          image: { modelId: this.metadata.mirroredModelId, imageName: this.imageName, imageTag: this.imageTag },
          ...this.logData,
        },
        'Completed registry upload',
      )

      resolve({
        metadata: this.metadata,
        image: { modelId: this.metadata.mirroredModelId, imageName: this.imageName, imageTag: this.imageTag },
      })
    } else {
      reject(
        InternalError('Manifest file (manifest.json) missing or invalid in Tarball file.', {
          metadata: this.metadata,
          ...this.logData,
        }),
      )
    }
  }
}
