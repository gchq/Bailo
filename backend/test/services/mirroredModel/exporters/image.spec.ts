import { beforeEach, describe, expect, test, vi } from 'vitest'

import { ReleaseAction } from '../../../../src/connectors/authorisation/actions.js'
import { ImageExporter } from '../../../../src/services/mirroredModel/exporters/image.js'
import { MirrorKind } from '../../../../src/services/mirroredModel/index.js'
import { Forbidden, InternalError } from '../../../../src/utils/error.js'

const tarballMocks = vi.hoisted(() => ({
  initialiseTarGzUpload: vi.fn(),
}))
vi.mock('../../../../src/services/mirroredModel/tarball.js', () => tarballMocks)

const mirroredModelMocks = vi.hoisted(() => ({
  addCompressedRegistryImageComponents: vi.fn(),
}))
vi.mock('../../../../src/services/mirroredModel/mirroredModel.js', () => mirroredModelMocks)

const registryMocks = vi.hoisted(() => ({
  joinDistributionPackageName: vi.fn(),
}))
vi.mock('../../../../src/services/registry.js', () => registryMocks)

const authMocks = vi.hoisted(() => ({
  default: {
    model: vi.fn(),
    release: vi.fn(),
    image: vi.fn(),
  },
}))
vi.mock('../../../../src/connectors/authorisation/index.js', () => authMocks)

const mockUser = { dn: 'userDN' } as any
const mockModel = {
  id: 'modelId',
  settings: { mirror: { destinationModelId: 'destModelId' } },
  card: { schemaId: 'schemaId' },
} as any
const mockImage = { _id: { toString: () => 'imageId' }, name: 'modelId/image', tag: 'latest' } as any
const mockRelease = { semver: '1.0.0', images: [mockImage] } as any
const mockLogData = { exporterType: 'ImageExporter', exportId: 'exportId' }

describe('connectors > mirroredModel > exporters > ImageExporter', () => {
  beforeEach(() => {
    tarballMocks.initialiseTarGzUpload.mockResolvedValue({
      tarStream: {} as any,
      gzipStream: {} as any,
      uploadStream: {} as any,
      uploadPromise: Promise.resolve(),
    })
    mirroredModelMocks.addCompressedRegistryImageComponents.mockResolvedValue(undefined)
    registryMocks.joinDistributionPackageName.mockReturnValue('joined/name:tag')
    authMocks.default.model.mockResolvedValue({ success: true })
    authMocks.default.release.mockResolvedValue({ success: true })
    authMocks.default.image.mockResolvedValue({ success: true })
  })

  test('constructor > sets release, image and logData', () => {
    const exporter = new ImageExporter(mockUser, mockModel, mockRelease, mockImage, mockLogData)

    expect(exporter.getRelease()).toEqual(mockRelease)
    expect(exporter.getImage()).toEqual(mockImage)
    expect(exporter).toMatchSnapshot()
  })

  test('_init > success sets distributionPackageName', async () => {
    const exporter = new ImageExporter(mockUser, mockModel, mockRelease, mockImage, mockLogData)

    // @ts-expect-error protected method
    await exporter._init()

    expect(authMocks.default.release).toHaveBeenCalledWith(mockUser, mockModel, ReleaseAction.View, mockRelease)
    expect(authMocks.default.image).toHaveBeenCalled()
    expect(exporter.getDistributionPackageName()).toBe('joined/name:tag')
  })

  test('_init > throws Forbidden if release auth fails', async () => {
    authMocks.default.release.mockResolvedValueOnce({ success: false, info: 'no release access' })
    const exporter = new ImageExporter(mockUser, mockModel, mockRelease, mockImage, mockLogData)

    // @ts-expect-error protected method
    await expect(exporter._init()).rejects.toEqual(
      Forbidden('no release access', {
        userDn: mockUser.dn,
        modelId: mockModel.id,
        semver: mockRelease.semver,
        ...mockLogData,
      }),
    )
  })

  test('_init > throws InternalError if image not in release', async () => {
    const exporter = new ImageExporter(mockUser, mockModel, { ...mockRelease, images: [] }, mockImage, mockLogData)

    await expect(
      // @ts-expect-error protected method
      exporter._init(),
    ).rejects.toEqual(
      InternalError('Could not find image associated with release.', {
        modelId: mockModel.id,
        semver: mockRelease.semver,
        imageId: mockImage._id.toString(),
        ...mockLogData,
      }),
    )
  })

  test('_init > throws Forbidden if image auth fails', async () => {
    authMocks.default.image.mockResolvedValueOnce({ success: false, info: 'no image access' })
    const exporter = new ImageExporter(mockUser, mockModel, mockRelease, mockImage, mockLogData)

    // @ts-expect-error protected method
    await expect(exporter._init()).rejects.toEqual(
      Forbidden('no image access', {
        userDn: mockUser.dn,
        modelId: mockModel.id,
        semver: mockRelease.semver,
        imageId: mockImage._id.toString(),
        ...mockLogData,
      }),
    )
  })

  test('getInitialiseTarGzUploadParams > returns correct params', () => {
    const exporter = new ImageExporter(mockUser, mockModel, mockRelease, mockImage, mockLogData)
    exporter['distributionPackageName'] = 'joined/name:tag'

    // @ts-expect-error protected method
    const params = exporter.getInitialiseTarGzUploadParams()

    expect(params[0]).toBe(`${mockImage._id.toString()}.tar.gz`)
    expect(params[1]).toMatchObject({
      exporter: mockUser.dn,
      sourceModelId: mockModel.id,
      mirroredModelId: mockModel.settings.mirror.destinationModelId,
      distributionPackageName: 'joined/name:tag',
      importKind: MirrorKind.Image,
    })
    expect(params[2]).toEqual(mockLogData)
  })

  test('getInitialiseTarGzUploadParams > throws if model missing', () => {
    const exporter = new ImageExporter(mockUser, undefined as any, mockRelease, mockImage, mockLogData)
    exporter['distributionPackageName'] = 'joined/name:tag'

    // @ts-expect-error protected method
    expect(() => exporter.getInitialiseTarGzUploadParams()).toThrowError(
      InternalError('Method `getInitialiseTarGzUploadParams` called before `this.model` defined.', mockLogData),
    )
  })

  test('getInitialiseTarGzUploadParams > throws if distributionPackageName missing', () => {
    const exporter = new ImageExporter(mockUser, mockModel, mockRelease, mockImage, mockLogData)

    // @ts-expect-error protected method
    expect(() => exporter.getInitialiseTarGzUploadParams()).toThrowError(
      InternalError(
        'Method `getInitialiseTarGzUploadParams` called before `this.distributionPackageName` defined.',
        mockLogData,
      ),
    )
  })

  test('addData > throws if not initialised', () => {
    const exporter = new ImageExporter(mockUser, mockModel, mockRelease, mockImage, mockLogData)

    expect(() => exporter.addData()).toThrowError(
      InternalError('Method `ImageExporter.addData` called before `init()`.', mockLogData),
    )
  })

  test('addData > success calls addCompressedRegistryImageComponents', async () => {
    const exporter = new ImageExporter(mockUser, mockModel, mockRelease, mockImage, mockLogData)
    exporter['initialised'] = true
    exporter['distributionPackageName'] = 'joined/name:tag'
    exporter['tarStream'] = {} as any
    exporter['gzipStream'] = {} as any
    exporter['uploadStream'] = {} as any
    exporter['uploadPromise'] = Promise.resolve()

    await exporter.addData()

    expect(mirroredModelMocks.addCompressedRegistryImageComponents).toHaveBeenCalledWith(
      mockUser,
      mockModel.id,
      'joined/name:tag',
      exporter['tarStream'],
      mockLogData,
    )
  })
})
