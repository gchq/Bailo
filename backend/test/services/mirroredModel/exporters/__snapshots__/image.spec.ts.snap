// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`connectors > mirroredModel > exporters > ImageExporter > constructor > sets release, image and logData 1`] = `
ImageExporter {
  "_init": [Function],
  "addData": [Function],
  "distributionPackageName": undefined,
  "getDistributionPackageName": [Function],
  "getImage": [Function],
  "getInitialiseTarGzUploadParams": [Function],
  "getRelease": [Function],
  "gzipStream": undefined,
  "image": {
    "_id": {
      "toString": [Function],
    },
    "name": "modelId/image",
    "tag": "latest",
  },
  "initialised": false,
  "logData": {
    "exportId": "exportId",
    "exporterType": "ImageExporter",
  },
  "model": {
    "card": {
      "schemaId": "schemaId",
    },
    "id": "modelId",
    "settings": {
      "mirror": {
        "destinationModelId": "destModelId",
      },
    },
  },
  "release": {
    "images": [
      {
        "_id": {
          "toString": [Function],
        },
        "name": "modelId/image",
        "tag": "latest",
      },
    ],
    "semver": "1.0.0",
  },
  "tarStream": undefined,
  "uploadPromise": undefined,
  "uploadStream": undefined,
  "user": {
    "dn": "userDN",
  },
}
`;
